python/az/aro/azext_aro/custom.py

```diff
@@ -338,3 +301,92 @@ def get_network_resources(cli_ctx, subnets, vnet):
     resources.update(route_tables)
 
     return resources
+
+
+# service_principal_update manages cluster service principal update
+# 1. If called without parameters it should be best-effort
+# 2. If called with parameters it fails if something is not possible
+# 2.1. If client_id is provided, we expect expect secret to be provided too
+#      If only secret is provided - we are updating secret
+#      In any case we will try to verify and update rbac
+# 2.2. TODO: refresh-cluster-service-principal is provided
+def service_principal_update(cli_ctx, oc, client_id=None, client_secret=None):
```

Comment:
I am interested to see how this can be factored in such a way as to share code with the aro_create path.  I wonder if the work of making the AAD calls will need to be more separate from the code to fix up the network resources, but I don't see it very clearly yet.

Reply:
I'm not sure how reviewable this is before --refresh-cluster-service-principal is implemented in the PR as well.